overlord/servicestate/quota_control_test.go

```diff
@@ -0,0 +1,322 @@
+		},
+	})
+	defer r()
+
+	// trying to create a quota with a snap that doesn't exist fails
+
+	err := s.mgr.CreateQuota("foo", "", []string{"test-snap"}, quantity.SizeGiB)
+	c.Assert(err, ErrorMatches, `cannot use snap "test-snap" in group "foo": snap "test-snap" is not installed`)
+
+	st := s.state
+	st.Lock()
+	// setup the snap so it exists
+	snapstate.Set(s.state, "test-snap", s.testSnapState)
+	snaptest.MockSnapCurrent(c, testYaml, s.testSnapSideInfo)
+	st.Unlock()
+
+	// now we can create the quota group
+	err = s.mgr.CreateQuota("foo", "", []string{"test-snap"}, quantity.SizeGiB)
+	c.Assert(err, IsNil)
+
+	// we can't add the same snap to a different group though
+	err = s.mgr.CreateQuota("foo2", "", []string{"test-snap"}, quantity.SizeGiB)
+	c.Assert(err, ErrorMatches, `cannot use snap "test-snap" in group "foo2": snap already in quota group "foo"`)
+
+	// creating the same group again will fail
+	err = s.mgr.CreateQuota("foo", "", []string{"test-snap"}, quantity.SizeGiB)
+	c.Assert(err, ErrorMatches, `group "foo" already exists`)
+
+	// check that the quota groups were created in the state
+	st.Lock()
+	defer st.Unlock()
+	m, err := servicestate.AllQuotas(st)
+	c.Assert(err, IsNil)
+	c.Assert(m, HasLen, 1)
+	for name, grp := range m {
+		switch name {
+		case "foo":
+			c.Assert(grp.Snaps, DeepEquals, []string{"test-snap"})
+			c.Assert(grp.SubGroups, HasLen, 0)
+			c.Assert(grp.ParentGroup, Equals, "")
+		default:
+			c.Errorf("unexpected group %q in state", name)
+		}
+	}
+}
+
+func (s *quotaControlSuite) TestCreateSubGroupQuota(c *C) {
+	r := s.mockSystemctlCalls(c, []expectedSystemctl{
+		{
+			// called for new slice unit written by CreateQuota after we create
+			// the snap in state
+			expArgs: []string{"daemon-reload"},
+		},
+	})
+	defer r()
+
+	st := s.state
+	st.Lock()
+	// setup the snap so it exists
+	snapstate.Set(s.state, "test-snap", s.testSnapState)
+	snaptest.MockSnapCurrent(c, testYaml, s.testSnapSideInfo)
+	st.Unlock()
+
+	// create a quota group with no snaps to be the parent
+	err := s.mgr.CreateQuota("foo", "", nil, quantity.SizeGiB)
+	c.Assert(err, IsNil)
+
+	// now we can create a sub-quota
+	err = s.mgr.CreateQuota("foo2", "foo", []string{"test-snap"}, quantity.SizeGiB)
+	c.Assert(err, IsNil)
+
+	// check that the quota groups were created in the state
+	st.Lock()
+	defer st.Unlock()
+	m, err := servicestate.AllQuotas(st)
+	c.Assert(err, IsNil)
+	c.Assert(m, HasLen, 2)
+	for name, grp := range m {
+		switch name {
+		case "foo":
+			c.Assert(grp.Snaps, HasLen, 0)
+			c.Assert(grp.SubGroups, DeepEquals, []string{"foo2"})
+			c.Assert(grp.ParentGroup, Equals, "")
+		case "foo2":
+			c.Assert(grp.Snaps, DeepEquals, []string{"test-snap"})
+			c.Assert(grp.SubGroups, HasLen, 0)
+			c.Assert(grp.ParentGroup, Equals, "foo")
+		default:
+			c.Errorf("unexpected group %q in state", name)
+		}
+	}
+}
+
+func (s *quotaControlSuite) TestRemoveQuota(c *C) {
+	r := s.mockSystemctlCalls(c, []expectedSystemctl{
+		{
```

Comment:
that's quite a bit of daemon-reload, seems we want a TODO somewhere to see what we can do to avoid some of them?

Reply:
There is a todo or at least there was in my branch maybe it didn't make it here...

Reply:
well added more todos about this